### PR TITLE
Fix space assignment for zone outside boundary condition and add space option

### LINKED ISSUE #10697: Surface with  Outside Boundary Condition=Zone does not work correctly if surface is assigned to a Space
Issue overview
--------------
From user: Attached (240819_Base EEM.idf) is a modified version of the prototype building ASHRAE901_ApartmentMidRise_STD2022_NewYork.idf 

The most significant change to the file is that each zone now contains one space.

Many of the surfaces in this building have "ZONE" selected as the :outside_boundary_condition

This file has space listed as :outside_boundary_condition_objects.  The error file (also attached - 240819_Base EEM.err)

The error file gives a repeated warning that is incorrect:

   ** Warning ** GetSurfaceData: CAUTION -- Interspace surfaces are usually in different spaces
   ** ~~~ ** Surface=G SIWALL N1A, Space=FLOOR 1 G N1 APARTMENT, Zone=G N1 APARTMENT
   ** ~~~ ** Surface=iz-G SIWALL N1A, Space=FLOOR 1 G N1 APARTMENT, Zone=G CORRIDOR

In 240824_Base EEM.idf - each surface is assigned both to the Zone and to the Space.  In 240824_Zone_Base EEM.idf each surface is assigned just to the Zone.  There are significant differences in the heating and cooling energy use just from making this change, which does not seem appropriate.

The results from 240824_Zone_Base EEM.idf more closely match the results from the original file.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Helpdesk ticket number 16745 and 16746

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added to EnergyPlusDevSupport\DefectFiles
 - [ ] Ticket added to EnergyPlus Defect Complexity (Github Project)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
 - Fixes #10697 
 - Currently, if a surface has been assigned to a Zone and a Space, and the Outside Boundary Condition = Zone, the new auto-generated other side surface is assigned to the correct Zone, but it is assigned to the same Space as the original surface.
```
** Warning ** GetSurfaceData: CAUTION -- Interspace surfaces are usually in different spaces
** ~~~ ** Surface=G SIWALL N1A, Space=FLOOR 1 G N1 APARTMENT, Zone=G N1 APARTMENT
** ~~~ ** Surface=iz-G SIWALL N1A, Space=FLOOR 1 G N1 APARTMENT, Zone=G CORRIDOR
```
 - Fix this to assign the correct Space to the new surface, but this will only work if the other side Zone has only one Space. If it has multiples spaces, then issue a fatal error.
 - For `BuildingSurface:Detailed`, `Wall:Detailed`, `RoofCeiling:Detailed`, and `Floor:Detailed` add "Space" as a choice for Outside Boundary Condition, so the user can specify the adjacent Space name.
 - For `Wall:Interzone`, `Ceiling:Interzone`, and `Floor:Interzone` enforce the single-space requirement.
 - For `Window:Interzone`, `Door:Interzone`, and `GlazedDoor:Interzone`, fix the code to use the same Space as the base surface.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] ~If any diffs are expected, author must demonstrate they are justified using plots and descriptions~
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] ~If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label~
 - [ ] ~If structural output changes, add to output rules file and add OutputChange label~
 - [ ] ~If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies~

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by mjwitte** [comment]
Need to test that this works for cases where the surface is adjacent to the same zone/space, such as a middle floor that's multiplied.

**Comment by mjwitte** [comment]
One of the new unit tests fails with develop:
```
Note: Google Test filter = *SurfaceGeometry_ZoneOutsideBC*
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from EnergyPlusFixture
[ RUN      ] EnergyPlusFixture.SurfaceGeometry_ZoneOutsideBC_SpacesNotInput
[       OK ] EnergyPlusFixture.SurfaceGeometry_ZoneOutsideBC_SpacesNotInput (11426 ms)
[ RUN      ] EnergyPlusFixture.SurfaceGeometry_ZoneOutsideBC_SpacesInput
C:\MJWGit\EnergyPlus\tst\EnergyPlus\unit\Fixtures\EnergyPlusFixture.cc(211): error: Expected equality of these values:
  expected_string
    Which is: ""
  stream_str
    Which is: "   ** Warning ** GetSurfaceData: CAUTION -- Interspace surfaces are occurring in the same space(s).\n   **   ~~~   ** ...use Output:Diagnostics,DisplayExtraWarnings; to show more details on individual occurrences.\n   ** Warning ** CalculateZoneVolume: 2 zones are not fully enclosed. For more details use:  Output:Diagnostics,DisplayExtrawarnings; \n"
With diff:
@@ -1,1 +1,3 @@
-""
+   ** Warning ** GetSurfaceData: CAUTION -- Interspace surfaces are occurring in the same space(s).
+   **   ~~~   ** ...use Output:Diagnostics,DisplayExtraWarnings; to show more details on individual occurrences.
+   ** Warning ** CalculateZoneVolume: 2 zones are not fully enclosed. For more details use:  Output:Diagnostics,DisplayExtrawarnings; \n

C:\MJWGit\EnergyPlus\tst\EnergyPlus\unit\SurfaceGeometry.unit.cc(15173): error: Value of: compare_err_stream("")
  Actual: false
Expected: true
C:\MJWGit\EnergyPlus\tst\EnergyPlus\unit\SurfaceGeometry.unit.cc(15184): error: Expected equality of these values:
  state->dataHeatBal->space(1).surfaces.size()
    Which is: 9
  7
C:\MJWGit\EnergyPlus\tst\EnergyPlus\unit\SurfaceGeometry.unit.cc(15185): error: Expected equality of these values:
  state->dataHeatBal->space(2).surfaces.size()
    Which is: 5
  7
C:\MJWGit\EnergyPlus\tst\EnergyPlus\unit\SurfaceGeometry.unit.cc(15204): error: Expected equality of these values:
  izIntWall.spaceNum
    Which is: 1
  2
C:\MJWGit\EnergyPlus\tst\EnergyPlus\unit\SurfaceGeometry.unit.cc(15217): error: Expected equality of these values:
  izIntDoor.spaceNum
    Which is: 1
  2
[  FAILED  ] EnergyPlusFixture.SurfaceGeometry_ZoneOutsideBC_SpacesInput (4696 ms)
[----------] 2 tests from EnergyPlusFixture (16126 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (16128 ms total)
[  PASSED  ] 1 test.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] EnergyPlusFixture.SurfaceGeometry_ZoneOutsideBC_SpacesInput

 1 FAILED TEST
```

**Comment by mjwitte** [comment]
*Results for the defect files.*

**240824_Zone_Base EEM** 
(same with develop and this branch, surfaces not assigned to spaces, Zone outside boundary condition used for some surfaces)
image

**240824_Base EEM**
 (surfaces assigned to spaces, Zone outside boundary condition used for some surfaces)

With develop
image

errors
```
   ** Warning ** GetSurfaceData: CAUTION -- Interspace surfaces are usually in different spaces
   **   ~~~   ** Surface=G SIWALL N1A, Space=FLOOR 1 G N1 APARTMENT, Zone=G N1 APARTMENT
   **   ~~~   ** Surface=iz-G SIWALL N1A, Space=FLOOR 1 G N1 APARTMENT, Zone=G CORRIDOR
   ** Warning ** GetSurfaceData: CAUTION -- Interspace surfaces are usually in different spaces
   **   ~~~   ** Surface=G EIWALL N1A, Space=FLOOR 1 G N1 APARTMENT, Zone=G N1 APARTMENT
   **   ~~~   ** Surface=iz-G EIWALL N1A, Space=FLOOR 1 G N1 APARTMENT, Zone=G N2 APARTMENT
etc
```
Interzone surfaces show in different zones but in the same space
image

With this branch
Energy results match the "zone" version.
image

Surface errors are gone.
And interzone surfaces are in different spaces.
image
image

File: idd/Energy+.idd.in [hunk 1]
```diff
@@ -10197,6 +10198,7 @@ BuildingSurface:Detailed,
        \key Adiabatic
        \key Surface
        \key Zone
+       \key Space
```
Comment (mjwitte) [inline idd/Energy+.idd.in:10201]: New outside boundary condition choice for the `Detailed` surface types.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 2]
```diff
@@ -108,9 +108,11 @@ namespace SurfaceGeometry {
 
     static std::string const BlankString;
 
-    int constexpr UnenteredAdjacentZoneSurface = -998; // allows users to enter one zone surface ("Zone")
-                                                       // referencing another in adjacent zone
-    int constexpr UnreconciledZoneSurface = -999;      // interim value between entering surfaces ("Surface") and reconciling
+    int constexpr unenteredAdjacentSpaceSurface = -997; // allows users to enter one zone surface ("Space")
```
Comment (mjwitte) [inline src/EnergyPlus/SurfaceGeometry.cc:111]: New flag variable for Space outside BC.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 3]
```diff
@@ -108,9 +108,11 @@ namespace SurfaceGeometry {
...
+                                                        // referencing another in adjacent space
+    int constexpr unenteredAdjacentZoneSurface = -998;  // allows users to enter one zone surface ("Zone")
+                                                        // referencing another in adjacent zone
+    int constexpr unreconciledZoneSurface = -999;       // interim value between entering surfaces ("Surface") and reconciling
```
Comment (mjwitte) [inline src/EnergyPlus/SurfaceGeometry.cc:115]: Somewhat gratuitous name changes to lower case to help reveal where these flags get used.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 4]
```diff
@@ -1195,125 +1194,115 @@ namespace SurfaceGeometry {
         // Debug    write(outputfiledebug,*) ' need to add ',NeedtoAddSurfaces+NeedToAddSubSurfaces
         for (int SurfNum = 1; SurfNum <= FirstTotalSurfaces; ++SurfNum) {
             auto &surfTemp = state.dataSurfaceGeometry->SurfaceTmp(SurfNum);
-            if (surfTemp.ExtBoundCond != UnenteredAdjacentZoneSurface) continue;
+            if ((surfTemp.ExtBoundCond != unenteredAdjacentZoneSurface) && (surfTemp.ExtBoundCond != unenteredAdjacentSpaceSurface)) continue;
```
Comment (mjwitte) [inline src/EnergyPlus/SurfaceGeometry.cc:1197]: For both the zone and space BC, need to add the new companion surface.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 5]
```diff
@@ -1195,125 +1194,115 @@ namespace SurfaceGeometry {
...
             // Need to add surface
             ++CurNewSurf;
             // Debug    write(outputfiledebug,*) ' adding surface=',curnewsurf
-            state.dataSurfaceGeometry->SurfaceTmp(CurNewSurf) = state.dataSurfaceGeometry->SurfaceTmp(SurfNum);
+            auto &newSurf = state.dataSurfaceGeometry->SurfaceTmp(CurNewSurf);
```
Comment (mjwitte) [inline src/EnergyPlus/SurfaceGeometry.cc:1201]: Lots of diffs here just due to the new shortcut for `newSurf`.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 6]
```diff
@@ -1195,125 +1194,115 @@ namespace SurfaceGeometry {
...
-            state.dataSurfaceGeometry->SurfaceTmp(CurNewSurf).Zone = Found;
-            state.dataSurfaceGeometry->SurfaceTmp(CurNewSurf).ZoneName = state.dataHeatBal->Zone(Found).Name;
+            if (surfTemp.ExtBoundCond == unenteredAdjacentZoneSurface) {
+                int Found = Util::FindItemInList(surfTemp.ExtBoundCondName, state.dataHeatBal->Zone, state.dataGlobal->NumOfZones);
+                if (Found == 0) continue;
+                newSurf.Zone = Found;
+                auto &newZone = state.dataHeatBal->Zone(Found);
+                newSurf.ZoneName = newZone.Name;
+                assert(newZone.spaceIndexes.size() >= 1);
+                newSurf.spaceNum = 0; // clear this here and set later
```
Comment (mjwitte) [inline src/EnergyPlus/SurfaceGeometry.cc:1211]: For the zone BC, set `spaceNum=0` and figure it out later.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 7]
```diff
@@ -1195,125 +1194,115 @@ namespace SurfaceGeometry {
...
+            } else if (surfTemp.ExtBoundCond == unenteredAdjacentSpaceSurface) {
+                int Found = Util::FindItemInList(surfTemp.ExtBoundCondName, state.dataHeatBal->space, state.dataGlobal->numSpaces);
+                if (Found == 0) continue;
+                newSurf.spaceNum = Found;
+                int zoneNum = state.dataHeatBal->space(Found).zoneNum;
+                newSurf.Zone = zoneNum;
+                newSurf.ZoneName = state.dataHeatBal->Zone(zoneNum).Name;
```
Comment (mjwitte) [inline src/EnergyPlus/SurfaceGeometry.cc:1218]: For space BC, set `spaceNum` here and locate the zone for that space.
The remaining changes in this function are just from the `newSurf` shortcut.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 8]
```diff
@@ -2721,7 +2713,10 @@ namespace SurfaceGeometry {
             }
             if (thisSurf.spaceNum > 0) {
                 anySurfacesWithSpace(thisSurf.Zone) = true;
-            } else {
+            } else if (thisSurf.ExtBoundCond != unreconciledZoneSurface) {
+                anySurfacesWithoutSpace(thisSurf.Zone) = true;
+            } else if (thisSurf.Name.substr(0, 3) != "iz-") {
+                // Only trigger a new space if the spaceless surface is not an autogenerated interzone surface
                 anySurfacesWithoutSpace(thisSurf.Zone) = true;
```
Comment (mjwitte) [inline src/EnergyPlus/SurfaceGeometry.cc:2720]: Without breaking other situations where a surface has no space assigned, set the appropriate flags so that an autogenerated surface with the Zone BC gets assigned to an appropriate space.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 9]
```diff
@@ -2761,6 +2757,21 @@ namespace SurfaceGeometry {
...
+                                                RoutineName,
+                                                thisSurf.Name.substr(3),
+                                                thisSurf.ZoneName));
+                        ShowContinueError(state,
+                                          format("Auto-generated surface=\"{}\" will be assigned to Space=\"{}\"",
+                                                 thisSurf.Name,
+                                                 state.dataHeatBal->space(thisSurf.spaceNum).Name));
+                        ShowContinueError(state, "Use Outside Boundary Condition = Space to specify the exact Space for the outside boundary.");
+                    }
+                }
```
Comment (mjwitte) [inline src/EnergyPlus/SurfaceGeometry.cc:2774]: After all of the spaces have been set up, every surface is already assigned to a zone, but some or all surfaces may not be assigned to a space yet. The existing logic assigns the last space in the zone. That logic remains the same here.

But if the surface was autogenerated using the Zone BC, but the other side zone has more than one space, a warning is issued, because the desired space is ambiguous.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 10]
```diff
@@ -3858,6 +3869,26 @@ namespace SurfaceGeometry {
...
+                    // This is the code for an unmatched "other surface"
+                    // will be set up later.
+                    state.dataSurfaceGeometry->SurfaceTmp(SurfNum).ExtBoundCond = unenteredAdjacentSpaceSurface;
+                    // check OutsideFaceEnvironment for legal zone
+                    Found = Util::FindItemInList(
+                        state.dataSurfaceGeometry->SurfaceTmp(SurfNum).ExtBoundCondName, state.dataHeatBal->space, state.dataGlobal->numSpaces);
+                    ++NeedToAddSurfaces;
+
+                    if (Found == 0) {
+                        ShowSevereError(state,
```
Comment (mjwitte) [inline src/EnergyPlus/SurfaceGeometry.cc:3882]: Severe error is space BC is specified, but the other side space name is not found.

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc [hunk 11]
```diff
@@ -14591,3 +14591,1524 @@ TEST_F(EnergyPlusFixture, ExtSolarForShadingTest)
     EXPECT_FALSE(thisSG->SurfaceTmp(8).HeatTransSurf);
     EXPECT_TRUE(thisSG->SurfaceTmp(8).ExtSolar);
 }
+TEST_F(EnergyPlusFixture, SurfaceGeometry_ZoneOutsideBC_SpacesNotInput)
```
Comment (mjwitte) [inline tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:14594]: 3 new unit test to exercise various combinations of zone and space BC.

**Review by mjwitte — COMMENTED** [review]
Code walkthru

**Comment by mjwitte** [comment]
> Need to test that this works for cases where the surface is adjacent to the same zone/space, such as a middle floor that's multiplied. *(comment from tech call on Jan 22)*

- Tested with modified version of MultiStory.idf where all of the adiabatic surfaces in the Middle floor zones were set to use the Zone boundary condition using the same zone that contains the surface. This changes the results from the base file due to more surfaces in the model, but the result is the same for both develop and this branch. This also introduces this warning:
```
   ** Warning ** GetSurfaceData: CAUTION -- Interspace surfaces are occurring in the same space(s).
   **   ~~~   ** ...use Output:Diagnostics,DisplayExtraWarnings; to show more details on individual occurrences.
```

- Then added a space for Mid Center Space and assigned all surfaces to Mid Center Space, keeping the outside BC as Zone, and again with outside BC as Space (Mid Center Space). No change in results. So I think this is working as expected.

[TestMiddleFloorSameZoneSpace.zip](https://github.com/user-attachments/files/18679690/TestMiddleFloorSameZoneSpace.zip)

File: src/EnergyPlus/SurfaceGeometry.cc (hunk 2 above)
Comment (Myoldmopar) [reply src/EnergyPlus/SurfaceGeometry.cc:111]: Seems ripe for an enum instead of hardcoded values, but there may be a reason that isn't ideal here.  Either way this doesn't have to hold up for it.

File: src/EnergyPlus/SurfaceGeometry.cc (hunk 7 above)
Comment (Myoldmopar) [reply src/EnergyPlus/SurfaceGeometry.cc:1218]: `newSurf` shortcut is a worthwhile cleanup for sure.

This is a message to anyone, not just you.  If you want to do these renaming changes, but they are going to cause lots of code diffs, feel free to fix them all up quickly in their own branch, and we can get it merged quickly so you can then make your targeted changes in a more compact PR.

In this case, it's not even close to worth doing that, it's just a few dozen extra lines changed.  So just noting it.

Comment (Myoldmopar) [reply src/EnergyPlus/SurfaceGeometry.cc:1218]: And it would actually be nice if you could add a pattern match on GitHub's UI to ignore some changes.  You can already ignore whitespace-only changes, so the capability is somewhat in place.  If you could just say "ignore any changes that are purely from "state.dataSurfaceGeometry->SurfaceTmp(CurNewSurf)" to "newSurf", that would be incredible.

FWIW I think they are using the --ignore-all-space or whatever argument with Git Diff.  There are other diff arguments that could be leveraged (-G along with --pickaxe-regex).  OK I'm done I'm done.

File: src/EnergyPlus/SurfaceGeometry.cc (hunk 8 above)
Comment (Myoldmopar) [reply src/EnergyPlus/SurfaceGeometry.cc:2720]: So you don't need an `else` block anymore at all?

File: src/EnergyPlus/SurfaceGeometry.cc (hunk 9 above)
Comment (Myoldmopar) [reply src/EnergyPlus/SurfaceGeometry.cc:2774]: I think I'm tracking this.  And seems reasonable to assume that and emit a warning.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 12]
```diff
@@ -3748,7 +3759,7 @@ namespace SurfaceGeometry {
                 if (Util::SameString(s_ipsc->cAlphaArgs(ArgPointer), "Outdoors")) {
                     surfTemp.ExtBoundCond = DataSurfaces::ExternalEnvironment;
                 } else if (Util::SameString(s_ipsc->cAlphaArgs(ArgPointer), "Adiabatic")) {
-                    surfTemp.ExtBoundCond = UnreconciledZoneSurface;
+                    surfTemp.ExtBoundCond = unreconciledZoneSurface;
```
Comment (Myoldmopar) [inline src/EnergyPlus/SurfaceGeometry.cc:3762]: I will say this would look real nice as an enum instead...  

```c++
tempSurface.exteriorBoundaryCondition = BC::UnreconciledZoneSurface;
```

File: src/EnergyPlus/SurfaceGeometry.cc (hunk 10 above)
Comment (Myoldmopar) [reply src/EnergyPlus/SurfaceGeometry.cc:3882]: Yes, reasonable.

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc (hunk 11 above)
Comment (Myoldmopar) [reply tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:14594]: Excellence.

**Review by Myoldmopar — APPROVED** [review]
Minor comments, but this looks great.  I'll look locally regarding my comment about the now-deleted `else`. I'm assuming you have the logic in place there, and it just caught me off guard in review.  Anyway, I'll test locally and confirm things.

**Comment by Myoldmopar** [comment]
Yeah this looks good to go.  Merging.  Thanks @mjwitte !
